[CI] FrozenIndexInputTests testRandomReads failing

Reproduces locally with the provided seed.

It looks like it broke in 5656ea1efa5caa6cfd04e7deb41b9e272469108f

**Build scan:**
https://gradle-enterprise.elastic.co/s/kdszwizfrenm2/tests/:x-pack:plugin:searchable-snapshots:test/org.elasticsearch.xpack.searchablesnapshots.store.input.FrozenIndexInputTests/testRandomReads

**Reproduction line:**
`./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.xpack.searchablesnapshots.store.input.FrozenIndexInputTests.testRandomReads" -Dtests.seed=9C1F9009A091BEF9 -Dtests.locale=sv -Dtests.timezone=IET -Druntime.java=11`

**Applicable branches:**
master

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.searchablesnapshots.store.input.FrozenIndexInputTests&tests.test=testRandomReads

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=37, name=TEST-FrozenIndexInputTests#randomReadAndSlice[T#3], state=RUNNABLE, group=TGRP-FrozenIndexInputTests]


  Caused by: java.lang.AssertionError: org.elasticsearch.common.util.concurrent.UncategorizedExecutionException: Failed execution

    at __randomizedtesting.SeedInfo.seed([9C1F9009A091BEF9]:0)
    at org.elasticsearch.common.lucene.store.ESIndexInputTestCase$1.onFailure(ESIndexInputTestCase.java:123)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.onFailure(ThreadContext.java:725)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.lang.Thread.run(Thread.java:834)

    Caused by: org.elasticsearch.common.util.concurrent.UncategorizedExecutionException: Failed execution

      at org.elasticsearch.common.util.concurrent.FutureUtils.rethrowExecutionException(FutureUtils.java:80)
      at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:50)
      at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:26)
      at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:177)
      at org.elasticsearch.common.lucene.store.ESIndexInputTestCase$1.doRun(ESIndexInputTestCase.java:141)
      at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737)
      at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
      at java.lang.Thread.run(Thread.java:834)

      Caused by: java.util.concurrent.ExecutionException: java.io.IOException: failed to read data from cache

        at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:262)
        at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:249)
        at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:76)
        at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:45)
        at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:26)
        at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:177)
        at org.elasticsearch.common.lucene.store.ESIndexInputTestCase$1.doRun(ESIndexInputTestCase.java:141)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(Thread.java:834)

        Caused by: java.io.IOException: failed to read data from cache

          at org.elasticsearch.xpack.searchablesnapshots.store.input.MetadataCachingIndexInput.readDirectlyIfAlreadyClosed(MetadataCachingIndexInput.java:410)
          at org.elasticsearch.xpack.searchablesnapshots.store.input.MetadataCachingIndexInput.doReadInternal(MetadataCachingIndexInput.java:111)
          at org.elasticsearch.xpack.searchablesnapshots.store.input.BaseSearchableSnapshotIndexInput.readInternal(BaseSearchableSnapshotIndexInput.java:112)
          at org.apache.lucene.store.BufferedIndexInput.refill(BufferedIndexInput.java:291)
          at org.apache.lucene.store.BufferedIndexInput.readByte(BufferedIndexInput.java:55)
          at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:66)
          at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.lambda$randomReadAndSlice$0(ESIndexInputTestCase.java:176)
          at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:434)
          at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:176)
          at org.elasticsearch.common.lucene.store.ESIndexInputTestCase$1.doRun(ESIndexInputTestCase.java:141)
          at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737)
          at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
          at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
          at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
          at java.lang.Thread.run(Thread.java:834)

          Caused by: java.lang.IndexOutOfBoundsException: Range [56889, 56889 + 1024) out of bounds for length 1024

            at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
            at jdk.internal.util.Preconditions.outOfBoundsCheckFromIndexSize(Preconditions.java:82)
            at jdk.internal.util.Preconditions.checkFromIndexSize(Preconditions.java:343)
            at java.util.Objects.checkFromIndexSize(Objects.java:424)
            at org.elasticsearch.common.bytes.BytesArray.slice(BytesArray.java:86)
            at org.elasticsearch.xpack.searchablesnapshots.store.input.MetadataCachingIndexInput.readWithBlobCache(MetadataCachingIndexInput.java:175)
            at org.elasticsearch.xpack.searchablesnapshots.store.input.MetadataCachingIndexInput.doReadInternal(MetadataCachingIndexInput.java:106)
            at org.elasticsearch.xpack.searchablesnapshots.store.input.BaseSearchableSnapshotIndexInput.readInternal(BaseSearchableSnapshotIndexInput.java:112)
            at org.apache.lucene.store.BufferedIndexInput.refill(BufferedIndexInput.java:291)
            at org.apache.lucene.store.BufferedIndexInput.readByte(BufferedIndexInput.java:55)
            at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:66)
            at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.lambda$randomReadAndSlice$0(ESIndexInputTestCase.java:176)
            at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:434)
            at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:176)
            at org.elasticsearch.common.lucene.store.ESIndexInputTestCase$1.doRun(ESIndexInputTestCase.java:141)
            at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737)
            at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
            at java.lang.Thread.run(Thread.java:834)

```

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheService.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheService.java
@@ -152,7 +152,7 @@ public class BlobStoreCacheService extends AbstractLifecycleComponent {
         }
     }
 
-    protected void getAsync(
+    final void getAsync(
         final String repository,
         final SnapshotId snapshotId,
         final IndexId indexId,
@@ -234,7 +234,7 @@ public class BlobStoreCacheService extends AbstractLifecycleComponent {
         return cause instanceof NodeClosedException || cause instanceof ConnectTransportException;
     }
 
-    public void putAsync(
+    public final void putAsync(
         final String repository,
         final SnapshotId snapshotId,
         final IndexId indexId,

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheService.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheService.java
@@ -167,7 +167,7 @@ public class BlobStoreCacheService extends AbstractLifecycleComponent {
             return;
         }
         final GetRequest request = new GetRequest(index).id(generateId(repository, snapshotId, indexId, shardId, name, range));
-        client.get(request, new ActionListener<>() {
+        innerGet(request, new ActionListener<>() {
             @Override
             public void onResponse(GetResponse response) {
                 if (response.isExists()) {
@@ -203,6 +203,10 @@ public class BlobStoreCacheService extends AbstractLifecycleComponent {
                 listener.onResponse(CachedBlob.CACHE_NOT_READY);
             }
         });
+    }
+
+    protected void innerGet(final GetRequest request, final ActionListener<GetResponse> listener) {
+        client.get(request, listener);
     }
 
     private static boolean assertDocId(
@@ -270,7 +274,7 @@ public class BlobStoreCacheService extends AbstractLifecycleComponent {
                     return;
                 }
                 final ActionListener<Void> wrappedListener = ActionListener.runAfter(listener, release);
-                client.index(request, new ActionListener<>() {
+                innerPut(request, new ActionListener<>() {
                     @Override
                     public void onResponse(IndexResponse indexResponse) {
                         logger.trace("cache fill ({}): [{}]", indexResponse.status(), request.id());
@@ -293,6 +297,10 @@ public class BlobStoreCacheService extends AbstractLifecycleComponent {
             logger.warn(() -> new ParameterizedMessage("cache fill failure: [{}]", id), e);
             listener.onFailure(e);
         }
+    }
+
+    protected void innerPut(final IndexRequest request, final ActionListener<IndexResponse> listener) {
+        client.index(request, listener);
     }
 
     protected static String generateId(

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/MetadataCachingIndexInput.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/MetadataCachingIndexInput.java
@@ -171,12 +171,8 @@ public abstract class MetadataCachingIndexInput extends BaseSearchableSnapshotIn
             assert bytesRead == length : bytesRead + " vs " + length;
         } else {
             final int sliceOffset = toIntBytes(position - cachedBlob.from());
-            assert sliceOffset + length <= cachedBlob.to() : "reading "
-                + length
-                + " bytes from "
-                + sliceOffset
-                + " exceed cached blob max position "
-                + cachedBlob.to();
+            assert sliceOffset + length <= cachedBlob.to()
+                : "reading " + length + " bytes from " + sliceOffset + " exceed cached blob max position " + cachedBlob.to();
 
             logger.trace("reading [{}] bytes of file [{}] at position [{}] using cache index", length, fileInfo.physicalName(), position);
             final BytesRefIterator cachedBytesIterator = cachedBlob.bytes().slice(sliceOffset, length).iterator();

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/MetadataCachingIndexInput.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/input/MetadataCachingIndexInput.java
@@ -170,8 +170,15 @@ public abstract class MetadataCachingIndexInput extends BaseSearchableSnapshotIn
             final int bytesRead = populateCacheFuture.get();
             assert bytesRead == length : bytesRead + " vs " + length;
         } else {
-            logger.trace("reading [{}] bytes of file [{}] at position [{}] using cache index", length, fileInfo.physicalName(), position);
             final int sliceOffset = toIntBytes(position - cachedBlob.from());
+            assert sliceOffset + length <= cachedBlob.to() : "reading "
+                + length
+                + " bytes from "
+                + sliceOffset
+                + " exceed cached blob max position "
+                + cachedBlob.to();
+
+            logger.trace("reading [{}] bytes of file [{}] at position [{}] using cache index", length, fileInfo.physicalName(), position);
             final BytesRefIterator cachedBytesIterator = cachedBlob.bytes().slice(sliceOffset, length).iterator();
             BytesRef bytesRef;
             int copiedBytes = 0;

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/TestUtils.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/TestUtils.java
@@ -9,8 +9,11 @@ package org.elasticsearch.xpack.searchablesnapshots.cache.common;
 import org.apache.lucene.mockfile.FilterFileChannel;
 import org.apache.lucene.mockfile.FilterFileSystemProvider;
 import org.apache.lucene.mockfile.FilterPath;
-import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.get.GetRequest;
+import org.elasticsearch.action.get.GetResponse;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.index.IndexResponse;
 import org.elasticsearch.client.Client;
 import org.elasticsearch.common.TriConsumer;
 import org.elasticsearch.common.blobstore.BlobContainer;
@@ -25,11 +28,10 @@ import org.elasticsearch.common.util.concurrent.DeterministicTaskQueue;
 import org.elasticsearch.core.CheckedConsumer;
 import org.elasticsearch.core.Nullable;
 import org.elasticsearch.core.PathUtilsForTesting;
+import org.elasticsearch.index.IndexNotFoundException;
+import org.elasticsearch.index.get.GetResult;
 import org.elasticsearch.index.shard.ShardId;
-import org.elasticsearch.repositories.IndexId;
-import org.elasticsearch.snapshots.SnapshotId;
 import org.elasticsearch.xpack.searchablesnapshots.cache.blob.BlobStoreCacheService;
-import org.elasticsearch.xpack.searchablesnapshots.cache.blob.CachedBlob;
 import org.elasticsearch.xpack.searchablesnapshots.store.IndexInputStats;
 
 import java.io.ByteArrayInputStream;
@@ -43,7 +45,6 @@ import java.nio.file.OpenOption;
 import java.nio.file.Path;
 import java.nio.file.attribute.FileAttribute;
 import java.nio.file.spi.FileSystemProvider;
-import java.time.Instant;
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
@@ -57,8 +58,11 @@ import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static java.util.Collections.synchronizedNavigableSet;
+import static org.elasticsearch.index.seqno.SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
+import static org.elasticsearch.index.seqno.SequenceNumbers.UNASSIGNED_SEQ_NO;
 import static org.elasticsearch.test.ESTestCase.between;
 import static org.elasticsearch.test.ESTestCase.randomLongBetween;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_BLOB_CACHE_INDEX;
 import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsUtils.toIntBytes;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.equalTo;
@@ -312,7 +316,7 @@ public final class TestUtils {
     public static class NoopBlobStoreCacheService extends BlobStoreCacheService {
 
         public NoopBlobStoreCacheService() {
-            super(null, mock(Client.class), null, () -> 0L);
+            super(null, mock(Client.class), SNAPSHOT_BLOB_CACHE_INDEX, () -> 0L);
         }
 
         @Override
@@ -321,44 +325,27 @@ public final class TestUtils {
         }
 
         @Override
-        protected void getAsync(
-            String repository,
-            SnapshotId snapshotId,
-            IndexId indexId,
-            ShardId shardId,
-            String name,
-            ByteRange range,
-            ActionListener<CachedBlob> listener
-        ) {
-            listener.onResponse(CachedBlob.CACHE_NOT_READY);
+        protected void innerGet(GetRequest request, ActionListener<GetResponse> listener) {
+            listener.onFailure(new IndexNotFoundException(request.index()));
+        }
+
+        @Override
+        protected void innerPut(IndexRequest request, ActionListener<IndexResponse> listener) {
+            listener.onFailure(new IndexNotFoundException(request.index()));
         }
 
         @Override
         public ByteRange computeBlobCacheByteRange(String fileName, long fileLength, ByteSizeValue maxMetadataLength) {
             return ByteRange.EMPTY;
         }
-
-        @Override
-        public void putAsync(
-            String repository,
-            SnapshotId snapshotId,
-            IndexId indexId,
-            ShardId shardId,
-            String name,
-            ByteRange range,
-            BytesReference bytes,
-            ActionListener<Void> listener
-        ) {
-            listener.onResponse(null);
-        }
     }
 
     public static class SimpleBlobStoreCacheService extends BlobStoreCacheService {
 
-        private final ConcurrentHashMap<String, CachedBlob> blobs = new ConcurrentHashMap<>();
+        private final ConcurrentHashMap<String, BytesArray> blobs = new ConcurrentHashMap<>();
 
         public SimpleBlobStoreCacheService() {
-            super(null, mock(Client.class), null, () -> 0L);
+            super(null, mock(Client.class), SNAPSHOT_BLOB_CACHE_INDEX, System::currentTimeMillis);
         }
 
         @Override
@@ -367,45 +354,38 @@ public final class TestUtils {
         }
 
         @Override
-        protected void getAsync(
-            String repository,
-            SnapshotId snapshotId,
-            IndexId indexId,
-            ShardId shardId,
-            String name,
-            ByteRange range,
-            ActionListener<CachedBlob> listener
-        ) {
-            CachedBlob blob = blobs.get(generateId(repository, snapshotId, indexId, shardId, name, range));
-            if (blob != null) {
-                listener.onResponse(blob);
-            } else {
-                listener.onResponse(CachedBlob.CACHE_MISS);
-            }
+        protected void innerGet(GetRequest request, ActionListener<GetResponse> listener) {
+            final BytesArray bytes = blobs.get(request.id());
+            listener.onResponse(
+                new GetResponse(
+                    new GetResult(
+                        request.index(),
+                        request.id(),
+                        UNASSIGNED_SEQ_NO,
+                        UNASSIGNED_PRIMARY_TERM,
+                        0L,
+                        bytes != null,
+                        bytes,
+                        null,
+                        null
+                    )
+                )
+            );
         }
 
         @Override
-        public void putAsync(
-            String repository,
-            SnapshotId snapshotId,
-            IndexId indexId,
-            ShardId shardId,
-            String name,
-            ByteRange range,
-            BytesReference bytes,
-            ActionListener<Void> listener
-        ) {
-            final CachedBlob cachedBlob = new CachedBlob(
-                Instant.ofEpochMilli(0),
-                Version.CURRENT,
-                repository,
-                name,
-                generatePath(snapshotId, indexId, shardId),
-                new BytesArray(bytes.toBytesRef(), true),
-                range.start()
+        protected void innerPut(IndexRequest request, ActionListener<IndexResponse> listener) {
+            final BytesArray bytesArray = blobs.put(request.id(), new BytesArray(request.source().toBytesRef(), true));
+            listener.onResponse(
+                new IndexResponse(
+                    new ShardId(request.index(), "_na", 0),
+                    request.id(),
+                    UNASSIGNED_SEQ_NO,
+                    UNASSIGNED_PRIMARY_TERM,
+                    0L,
+                    bytesArray == null
+                )
             );
-            blobs.put(generateId(repository, snapshotId, indexId, shardId, name, range), cachedBlob);
-            listener.onResponse(null);
         }
     }
 

